### PR TITLE
Fix handling union type values in the Avro payload

### PR DESCRIPTION
**Pull request by Nuvindu**

### CODE REVIEW

File: native/src/main/java/io/ballerina/lib/avro/serialize/visitor/SerializeVisitor.java [hunk 1]
```diff
@@ -102,6 +104,13 @@ public Object visit(PrimitiveSerializer primitiveSerializer, Object data) throws
             case FLOAT -> {
                 return ((Double) data).floatValue();
             }
+            case DOUBLE -> {
+                if (data instanceof Long) {
+                    return ((Long) data).doubleValue();
```
Comment (NipunaRanasinghe) [inline native/src/main/java/io/ballerina/lib/avro/serialize/visitor/SerializeVisitor.java]: Cant we do
```suggestion
                if (data instanceof Long longData) {
                    return longData.doubleValue();
```
?

File: native/src/main/java/io/ballerina/lib/avro/serialize/visitor/SerializeVisitor.java [hunk 2]
```diff
@@ -95,32 +98,32 @@ private Object serializeField(Schema schema, Object fieldData) throws Exception
...
-            case INT -> {
-                return ((Long) data).intValue();
-            }
-            case FLOAT -> {
-                return ((Double) data).floatValue();
+        return switch (primitiveSerializer.getSchema().getType()) {
+            case INT -> ((Long) data).intValue();
+            case FLOAT -> ((Double) data).floatValue();
+            case DOUBLE -> {
+                if (data instanceof Long longValue) {
```
Comment (SasinduDilshara) [inline native/src/main/java/io/ballerina/lib/avro/serialize/visitor/SerializeVisitor.java:115]: Lets add cases for `int`, `byte` types to map with ballerina integer subtypes.

File: ballerina/tests/record_tests.bal [hunk 3]
```diff
@@ -370,3 +370,105 @@ public isolated function testOptionalMultipleFieldsInRecords() returns error? {
...
+        currentGeoSegment: "HRD_FLORIDA",
+        kycStatus: "VERIFIED",
+        signupGeoSegment: "HRD_FLORIDA",
+        hrdSportsBetAmountLifetime: 731,
+        hrdOptInEmail: false,
+        hrdOptInPush: false
+    };
+    Schema avro = check new (schema);
+    byte[] serializedValue = check avro.toAvro(recordValue);
+    var deserializedValue = check avro.fromAvro(serializedValue, Record);
```
Comment (NipunaRanasinghe) [inline ballerina/tests/record_tests.bal]: any reason to use `var` here?

File: native/src/main/java/io/ballerina/lib/avro/serialize/visitor/SerializeVisitor.java [hunk 4]
```diff
@@ -97,39 +98,32 @@ private Object serializeField(Schema schema, Object fieldData) throws Exception
...
-        switch (primitiveSerializer.getSchema().getType()) {
-            case INT -> {
-                return ((Long) data).intValue();
-            }
-            case FLOAT -> {
-                return ((Double) data).floatValue();
-            }
+        return switch (primitiveSerializer.getSchema().getType()) {
+            case INT -> ((Long) data).intValue();
+            case FLOAT -> ((Double) data).floatValue();
```
Comment (NipunaRanasinghe) [inline native/src/main/java/io/ballerina/lib/avro/serialize/visitor/SerializeVisitor.java]: I'm not sure if these casts are always safe, as we try to cast based on the schema type in here. Shouldn't we add `instanceof` checks and do proper error handling in that case?

File: native/src/main/java/io/ballerina/lib/avro/serialize/visitor/SerializeVisitor.java (hunk 2 above)
Comment (Nuvindu) [reply native/src/main/java/io/ballerina/lib/avro/serialize/visitor/SerializeVisitor.java:115]: Added in [a05d61b](https://github.com/ballerina-platform/module-ballerina-avro/pull/27/commits/a05d61bfa3e4bb399001436e5df0cefe9bf2b20b)

**Review by NipunaRanasinghe — APPROVED** [review]

**Comment by NipunaRanasinghe** [comment]
@Nuvindu Lets update the changelog as well

**Pull request merged**